src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (in1_op == Op_Mul(bt) && (in1->in(1)->is_Con() || in1->in(2)->is_Con())) {
+    Node* base = in1->in(1)->is_Con() ? in1->in(2) : in1->in(1);
+    if (base != in2) {
+      return nullptr;
+    }
+
+    Node* con = in1->in(1)->is_Con() ? in1->in(1) : in1->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
+      return nullptr;
+    }
+
+    multiplier = con->get_integer_as_long(con_bt);
+    matched = true;
+  }
+
+  // Due to potential power-of-2 multiplication optimization, e.g., ((a << CON1) + (a << CON2)) + a,
+  // we also look for LHS patterns:
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  if (in1_op == Op_Add(bt) && in1->isa_Add()->is_optimized_multiplication()) {
+    Node* lhs = in1->in(1);
+    Node* rhs = in1->in(2);
+
+    Node* lhs_base = lhs->is_LShift() ? lhs->in(1) : lhs;
+    Node* rhs_base = rhs->is_LShift() ? rhs->in(1) : rhs;
+
+    Node* nodes[] = {lhs, rhs};
+    Node* bases[] = {lhs_base, rhs_base};
+    jlong multipliers[] = {1, 1};
+
+    // AddNode(LShiftNode(a, CON1), LShiftNode(a, CON2)/a)
+    // AddNode(LShiftNode(a, CON1)/a, LShiftNode(a, CON2))
+    for (int i = 0; i < 2; i++) {
+      if (nodes[i]->is_LShift()) {
+        Node* con = nodes[i]->in(2);
+        if (bases[i] != in2 || bases[0] != bases[1] || !con->is_Con()) {
+          return nullptr;
+        }
+
+        BasicType con_bt = phase->type(con)->basic_type();
+        if (con_bt == T_VOID) {
+          return nullptr;
+        }
+
+        multipliers[i] = ((jlong) 1 << con->get_integer_as_long(con_bt));
+      }
+    }
+
+    multiplier = multipliers[0] + multipliers[1];
+    matched = true;
+  }
+
+  // Due to potential power-of-2 multiplication optimization, e.g., ((a << CON) - a) + a,
+  //     - SubNode(LShiftNode(a, CON), a)
+  if (in1_op == Op_Sub(bt) && in1->in(1)->is_LShift() && in1->in(1)->in(2)->is_Con()) {
+    Node* lshift = in1->in(1);
+    Node* base = in1->in(2);
+
+    Node* lshift_base = lshift->in(1);
+    if (lshift_base != base || lshift_base != in2) {
+      return nullptr;
+    }
+
+    Node* con = lshift->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
+      return nullptr;
+    }
+
+    // We can't simply return the lshift node even if ((a << CON) - a) + a cancels out. Ideal() must return a new node.
+    multiplier = ((jlong) 1 << con->get_integer_as_long(con_bt)) - 1;
+    matched = true;
+  }
+
+  if (!matched) {
+    return nullptr;
+  }
+
+  multiplier++; // + 1 for current addition
+  Node* con = (bt == T_INT)
+              ? (Node*) phase->intcon((jint) multiplier) // intentional type narrowing to allow overflow at max_jint
+              : (Node*) phase->longcon(multiplier);
+  Node* mul = MulNode::make(con, in2, bt);
+
+  PhaseIterGVN* igvn = phase->is_IterGVN();
+  if (igvn != nullptr) {
```

Comment:
Why do you need that?
I think it's fine to return a new node from Ideal.

Reply:
You are right. This is leftover code from last version. Removed.